keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1030 @@
+
+//TODO update figure
+![proposal](proposal2.JPG)
+
+#### Device Attribute Additions
+
+To indicate whether a device is a fabric device, an attribute is added to the `Basic` within `Device`.
+This attribute will be used by the controller that exposes the `ResourceSlice` to notify whether the device is a fabric device.
+To avoid impacting existing DRA functionality, the default value of this attribute is set to `false`.
+
+```go
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+type Device struct {
+    // Name is unique identifier among all devices managed by
+    // the driver in the pool. It must be a DNS label.
+    //
+    // +required
+    Name string
+
+    // Basic defines one device instance.
+    //
+    // +optional
+    // +oneOf=deviceType
+    Basic *BasicDevice
+}
+
+// BasicDevice represents a basic device instance.
+type BasicDevice struct {
+    // Attributes contains additional attributes of the device.
+    //
+    // +optional
+    Attributes map[string]string
+}
+```
+
+To indicate a fabric device, the following attribute will be added:
+
+```yaml
+attributes:
+  kubernetes.io/needs-attaching:
+    bool: "true"
+```
+
+#### `AllocatedDeviceStatus` Additions
+
+The `Conditions` field within `AllocatedDeviceStatus` is used to indicate the status of the device attachment.
+This field will contain a list of conditions, each representing a specific state or event related to the device.
+
+For this feature, the NodeName and following `ConditionType` constants are added:
+
+```go
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+...
+    // NodeName contains the name of the node where the device needs to be attached.
+    //
+    // +optional
+    NodeName string 
+}
+const(
+  DRADeviceNeedAttachType = "kubernetes.io/needs-attaching"
+  DRADeviceIsAttachType = "kubernetes.io/is-attached"
+  DRADeviceAttachFailType = "kubernetes.io/attach-failed"
+)
+```
+
+When `kubernetes.io/needs-attaching: true` is set, the scheduler DRA plugin is expected to do the following:
+
+1. Set `AllocatedDeviceStatus.NodeName`.
+2. Add an `AllocatedDeviceStatus` with a condition of `Type: kubernetes.io/needs-attaching` and `Status: True`.
+3. Wait for a condition with `Type: kubernetes.io/is-attached` and `Status: True` in `PreBind` before proceeding.
+4. Give up when seeing a condition with `Type: kubernetes.io/attach-failed` and `Status: True`.
+
+Note: There is a concern that the in-flight events cache may grow too large when waiting in PreBind.
+This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.
+
+#### PreBind Phase Timeout
+
+If the device attachment is successful, we expect it to take no longer than 5 minutes.
```

Comment:
Have a question regarding the device attach timeout, do we expect backoff if it timeout consistently?

Reply:
Thank you for your question regarding the device attach timeout. At this stage, backoff is one of the options we are considering if the timeout occurs consistently. However, we have not yet finalized the details. Additionally, if a timeout occurs, it might be better to consider all attempts as failures and reschedule them.

Reply:
Got it, thanks